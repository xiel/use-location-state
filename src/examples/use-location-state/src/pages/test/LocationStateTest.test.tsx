import React from 'react'
import { act, cleanup, fireEvent, render } from '@testing-library/react'
import LocationStateDemo from '../LocationStateDemo'

afterEach(() => {
  // Restore all spyOn mocks
  jest.restoreAllMocks()

  window.history.replaceState(null, '', '/')
  cleanup()
})

describe('LocationStateDemo', () => {
  test('LocationStateDemo renders without crash/loop', () => {
    expect(window.history.state).toBe(null)
    expect(render(<LocationStateDemo />))
  })

  test('can set name with button', () => {
    expect(window.history.state).toBe(null)
    const { getByText, getByTestId } = render(<LocationStateDemo />)

    // should put new names into the hash (and age default value comes along)
    act(() => void fireEvent.click(getByText('name: "Felix"')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()

    act(() => void fireEvent.click(getByText('name: "Kim"')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()

    // set back to default value, so it should remove name value from hash
    act(() => void fireEvent.click(getByText('name: "Sarah" (default value)')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set name with text field', () => {
    expect(window.history.state).toBe(null)
    const { getByLabelText, getByTestId } = render(<LocationStateDemo />)
    fireEvent.change(getByLabelText('name:'), { target: { value: 'Mila' } })
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set age with button', () => {
    expect(window.history.state).toBe(null)
    const { getByText, getByTestId } = render(<LocationStateDemo />)
    act(() => void fireEvent.click(getByText('age: 30')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
    act(() => void fireEvent.click(getByText('age: 45')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
    act(() => void fireEvent.click(getByText('age: 25')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set age with text field', () => {
    expect(window.history.state).toBe(null)
    const { getByLabelText, getByTestId } = render(<LocationStateDemo />)
    fireEvent.change(getByLabelText('age:'), { target: { value: '33' } })
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set name & age with button', () => {
    expect(window.history.state).toBe(null)
    const { getByText, getByTestId } = render(<LocationStateDemo />)

    act(() => void fireEvent.click(getByText('name: "Felix", age: 30')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
    act(() => void fireEvent.click(getByText('name: "Kim", age: 45')))
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
    // set back to default value
    act(
      () => void fireEvent.click(getByText('name: "Sarah", age: 25 (default)'))
    )
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set active with checkbox', () => {
    expect(window.history.state).toBe(null)
    const { getByLabelText, getByTestId } = render(<LocationStateDemo />)
    act(() => {
      fireEvent.click(getByLabelText('active'))
    })
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set active with checkbox - push', () => {
    const replaceState = jest.spyOn(window.history, 'replaceState')
    const pushState = jest.spyOn(window.history, 'pushState')
    expect(window.history.state).toBe(null)
    const { getByLabelText, getByTestId } = render(<LocationStateDemo />)
    expect(replaceState).toBeCalledTimes(0)
    expect(pushState).toBeCalledTimes(0)
    fireEvent.click(getByLabelText('active (method: push)'))
    expect(replaceState).toBeCalledTimes(0)
    expect(pushState).toBeCalledTimes(1)
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set date with date field', () => {
    expect(window.history.state).toBe(null)
    const { getByLabelText, getByTestId } = render(<LocationStateDemo />)
    act(() => {
      fireEvent.change(getByLabelText('date:'), {
        target: { value: '2019-05-01' },
      })
    })
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })

  test('can set null in date field', () => {
    expect(window.history.state).toBe(null)
    const { getByLabelText, getByTestId } = render(<LocationStateDemo />)
    act(() => {
      fireEvent.change(getByLabelText('date:'), {
        target: { value: null },
      })
    })
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
    act(() => {
      fireEvent.change(getByLabelText('date:'), {
        target: { value: '' },
      })
    })
    expect(getByTestId('pre-query-state')).toMatchSnapshot()
  })
})
