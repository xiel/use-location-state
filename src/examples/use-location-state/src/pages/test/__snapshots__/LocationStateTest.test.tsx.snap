// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationStateDemo can set active with checkbox - push 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": true,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set active with checkbox 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": true,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set age with button 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 30,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set age with button 2`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 45,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set age with button 3`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set age with text field 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 33,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set date with date field 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": false,
 "date": "2019-05-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name & age with button 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Felix",
 "age": 30,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name & age with button 2`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Kim",
 "age": 45,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name & age with button 3`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name with button 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Felix",
 "age": 25,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name with button 2`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Kim",
 "age": 25,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name with button 3`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set name with text field 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Mila",
 "age": 25,
 "active": false,
 "date": "2019-01-01T00:00:00.000Z"
}
</pre>
`;

exports[`LocationStateDemo can set null in date field 1`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": false,
 "date": null
}
</pre>
`;

exports[`LocationStateDemo can set null in date field 2`] = `
<pre
  data-testid="pre-query-state"
>
  {
 "name": "Sarah",
 "age": 25,
 "active": false,
 "date": null
}
</pre>
`;
